### ISSUE TITLE
Facets for nested mappings don't display any values

### ISSUE THREAD

**Issue by danxmoran** [DataBiosphere/data-explorer#175]
I indexed [this table](https://console.cloud.google.com/bigquery?folder=&organizationId=548622027621&project=broad-gdr-encode-storage&p=broad-gdr-encode-storage&d=encode_2018_10_06&t=files_with_donors_denormalized&page=table) using the BQ indexer, producing the mapping:
```json
{
  "encode": {
    "aliases": {},
    "mappings": {
      "type": {
        "properties": {
          "broad-gdr-encode-storage": {
            "properties": {
              "encode_2018_10_06": {
                "properties": {
                  "files_with_donors_denormalized": {
                    "properties": {
                      "assay_term_name": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "assembly": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "audit_color": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "biosample_accession": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "biosample_term_id": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "biosample_type": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "cell_type": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "derived_from_exp": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "derived_from_ref": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "donors": {
                        "type": "nested",
                        "properties": {
                          "accession": {
                            "type": "text",
                            "fields": {
                              "keyword": {
                                "type": "keyword",
                                "ignore_above": 256
                              }
                            }
                          },
                          "age": {
                            "type": "text",
                            "fields": {
                              "keyword": {
                                "type": "keyword",
                                "ignore_above": 256
                              }
                            }
                          },
                          "age_units": {
                            "type": "text",
                            "fields": {
                              "keyword": {
                                "type": "keyword",
                                "ignore_above": 256
                              }
                            }
                          },
                          "health_status": {
                            "type": "text",
                            "fields": {
                              "keyword": {
                                "type": "keyword",
                                "ignore_above": 256
                              }
                            }
                          },
                          "sex": {
                            "type": "text",
                            "fields": {
                              "keyword": {
                                "type": "keyword",
                                "ignore_above": 256
                              }
                            }
                          },
                          "url": {
                            "type": "text",
                            "fields": {
                              "keyword": {
                                "type": "keyword",
                                "ignore_above": 256
                              }
                            }
                          }
                        }
                      },
                      "experiment_accession": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "file_format": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "file_size": {
                        "type": "long"
                      },
                      "file_type": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "gcs_file_available": {
                        "type": "boolean"
                      },
                      "lab_name": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "library_accession": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "md5sum": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "output_type": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "path": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "percent_aligned": {
                        "type": "float"
                      },
                      "percent_duplicated": {
                        "type": "float"
                      },
                      "read_count": {
                        "type": "float"
                      },
                      "read_length": {
                        "type": "float"
                      },
                      "run_type": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "status": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "target": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "url": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "uuid": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      },
                      "warning_summary": {
                        "type": "text",
                        "fields": {
                          "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                          }
                        }
                      }
                    }
                  }
                }
              }
          }
        }
      }
    }
}
```
I then spun up the explorer API and UI using this config:
```json
{
    "enable_field_search": false,
    "facets": [
        {
            "ui_facet_name": "File Format",
            "elasticsearch_field_name": 
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.file_format"
        },
        {
            "ui_facet_name": "File Type",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.file_type"
        },
        {
            "ui_facet_name": "Donor Sex",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.donors.sex"
        },
        {
            "ui_facet_name": "Donor Health Status",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.donors.health_status"
        },
        {
            "ui_facet_name": "Sample Type",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.biosample_type"
        },
        {
            "ui_facet_name": "Sample Term ID",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.biosample_term_id"
        },
        {
            "ui_facet_name": "Run Type",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.run_type"
        },
        {
            "ui_facet_name": "Output Type",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.output_type"
        },
        {
            "ui_facet_name": "Cell Type",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.cell_type"
        },
        {
            "ui_facet_name": "Lab Name",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.lab_name"
        },
        {
            "ui_facet_name": "Target",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.target"
        },
        {
            "ui_facet_name": "File Assembly",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.assembly"
        },
        {
            "ui_facet_name": "File Status",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.status"
        },
        {
            "ui_facet_name": "Cloud File Available",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.gcs_file_available"
        },
        {
            "ui_facet_name": "Audit Level",
            "elasticsearch_field_name":
                "broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.audit_color"
        }
    ]
}
```
The result is running [here](http://broad-gdr-encode.appspot.com/). Facets for the "sex" and "health_status" fields of the nested "donors" mapping load, but show no possible values.

**Comment by melissachang** [comment]
Thanks for the detailed report Dan.

The problem happens with BigQuery columns that are REPEATED (which we haven't encountered before Encode).

Note that if the BigQuery `donors` column was RECORD, things would work. The `broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.donors.health_status` facet would work.

The problem happens because the BigQuery `donors` column is REPEATED. In Elasticsearch, the `donors` field is an array of objects. Additionally, the `donors` field [must be nested in order for faceted search to work](https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html).

Thankfully @bfcrampton [changed the indexer](https://github.com/DataBiosphere/data-explorer-indexers/blob/a21e47095d84162ca60cef881651e0381d6c684f/bigquery/indexer.py#L93) to make the Elasticsearch `donors` field nested -- see the `"type": "nested"` under `"donors"` in your mappings above. So the index is correct. What's left is, to fix the faceted search.

We [construct Elasticsearch facet objects here](https://github.com/DataBiosphere/data-explorer/blob/5db00e5fa38ca0326b6b5b1106ec0319febd6c3f/api/data_explorer/util/elasticsearch_util.py#L164). Currently we return `TermsFacet` for `broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.donors.health_status`. We need to somehow detect that `donors` is nested and return `NestedFacet` instead. `NestedFacet` takes in a `path`; in this case `path` is `broad-gdr-encode-storage.encode_2018_10_06.files_with_donors_denormalized.donors`.

Currently we special-case `samples` facets and return `NestedFacet`. As part of this fix, we should no longer need to special-case `samples` facets.

**Comment by melissachang** [comment]
Fixed by #180.
